tests/conftest.py

```diff
@@ -1,18 +1,42 @@
+import atexit
 import pytest
 import pytest_asyncio
 import tempfile
 from pathlib import Path
+from chia.consensus.constants import ConsensusConstants
+from tests.block_tools import BlockTools, test_constants, create_block_tools
+from tests.util.keyring import TempKeyring
+from typing import Tuple
 
 
-# TODO: tests.setup_nodes (which is also imported by tests.util.blockchain) creates a
-#       global BlockTools at tests.setup_nodes.bt.  This results in an attempt to create
-#       the chia root directory which the build scripts symlink to a sometimes-not-there
-#       directory.  When not there Python complains since, well, the symlink is a file
-#       not a directory and also not pointing to a directory.  In those same cases,
-#       these fixtures are not used.  It would be good to refactor that global state
-#       creation, including the filesystem modification, away from the import but
-#       that seems like a separate step and until then locating the imports in the
-#       fixtures avoids the issue.
+def cleanup_keyring(keyring: TempKeyring):
+    keyring.cleanup()
+
+
+temp_keyring = TempKeyring(populate=True)
+keychain = temp_keyring.get_keychain()
+atexit.register(cleanup_keyring, temp_keyring)  # Attempt to clean up the temp keychain
+
+
+@pytest.fixture(scope="session", name="bt")
+def bt() -> BlockTools:
+    _shared_block_tools = create_block_tools(constants=test_constants, keychain=keychain)
+    return _shared_block_tools
+    # yield _shared_block_tools
+    # _shared_block_tools.cleanup()
+
+
+@pytest.fixture(scope="session")
+def self_hostname(bt):
+    return bt.config["self_hostname"]
+
+
+def setup_shared_block_tools_and_keyring(
```

Comment:
I don't see this used anywhere. It also looks suspicious, now that `bt` is a fixture, it doesn't look right to also create one like this (that also may not be cleaned up correctly on shutdown).

Is it possible to *set* the keychain on an existing `bt` object? If so, tests that need a specific keyring could still use the `bt` fixture.

Reply:
That is a really good idea. I will add a comment.
However, I think the keychain requires that there is only one instance at the moment.